pkg/ds/daemon_set.go

```diff
@@ -0,0 +1,353 @@
+	ds.logger.NoFields().Infof("Need to unschedule %d nodes", len(toUnscheduleSorted))
+
+	for _, node := range toUnscheduleSorted {
+		err := ds.unschedule(node)
+		if err != nil {
+			return util.Errorf("Error scheduling node: %v", err)
+		}
+	}
+	return nil
+}
+
+// clearPods unschedules pods for all the nodes that have been scheduled by
+// this daemon set by using CurrentNodes()
+// This should only be used when a daemon set is deleted
+func (ds *daemonSet) clearPods() error {
+	currentNodes, err := ds.CurrentNodes()
+	if err != nil {
+		return util.Errorf("Error retrieving current nodes daemon set: %v", err)
+	}
+
+	// Get the difference in nodes that we need to unschedule on and then sort them
+	// for deterministic ordering
+	toUnscheduleSorted := types.NewNodeSet(currentNodes...).ListNodes()
+	ds.logger.NoFields().Infof("Need to unschedule %d nodes", len(toUnscheduleSorted))
+
+	for _, node := range toUnscheduleSorted {
+		err := ds.unschedule(node)
+		if err != nil {
+			return util.Errorf("Error scheduling node: %v", err)
+		}
+	}
+	return nil
+}
+
+func (ds *daemonSet) schedule(node types.NodeName) error {
+	ds.logger.NoFields().Infof("Scheduling on %s", node)
+
+	// Will apply the following label on the key <labels.NODE>/<node.String()>:
+	// 	{ DSIDLabel : ds.ID().String() }
+	// eg node/127.0.0.1[daemon_set_id] := test_ds_id
+	err := ds.nodeApplicator.SetLabel(labels.NODE, node.String(), DSIDLabel, ds.ID().String())
+	if err != nil {
+		return util.Errorf("Error setting label: %v", err)
+	}
+
+	// Will set the following the value, ds.Manifest, to the following key:
+	// <kp.INTENT_TREE>/<node>/<ds.Manifest.ID()>
+	// eg intent/127.0.0.1/test = test_pod
+	_, err = ds.kpStore.SetPod(kp.INTENT_TREE, node, ds.Manifest)
+	if err != nil {
+		return util.Errorf("Error adding pod to intent tree: %v", err)
+	}
+	return nil
+}
+
+func (ds *daemonSet) unschedule(node types.NodeName) error {
+	ds.logger.NoFields().Infof("Unscheduling from %s", node)
+
+	// Will remove the following key:
+	// <kp.INTENT_TREE>/<node>/<ds.Manifest.ID()>
+	_, err := ds.kpStore.DeletePod(kp.INTENT_TREE, node, ds.Manifest.ID())
+	if err != nil {
+		return util.Errorf("Unable to delete pod from intent tree: %v", err)
+	}
+
+	// Will remove the following label on the key <labels.NODE>/<node>: DSIDLabel
+	err = ds.nodeApplicator.RemoveLabel(labels.NODE, node.String(), DSIDLabel)
+	if err != nil {
+		return util.Errorf("Error removing label: %v", err)
+	}
+
+	return nil
+}
+
+func (ds *daemonSet) CurrentNodes() ([]types.NodeName, error) {
+	// Changing DaemonSet.ID is not permitted, so as long as there is no
+	// uuid collision, this will always get the current nodes that this daemon set
+	// had scheduled on
+	selector := klabels.Everything().Add(DSIDLabel, klabels.EqualsOperator, []string{ds.ID().String()})
+
+	nodeMatches, err := ds.nodeApplicator.GetMatches(selector, labels.NODE)
```

Comment:
`labels.POD` here too, replication controller likely has prior art for parsing the resulting strings since it's `<node>/<pod_id>`